[master] Fix detection of Salt codename by "salt_version" execution module.

meaksh opened the pull request:
### What does this PR do?

This PR fixes an issue and some kind of mess we have around Salt version detection and codename resolution.

Currently, the codename resolution (`salt_version.get_release_number` and `salt_version.equal`) seems broken in Salt:

```console
# pip3 install salt
Collecting salt
  Downloading salt-3006.1.tar.gz (20.4 MB)
[...]
Successfully installed Jinja2-3.0.3 MarkupSafe-2.0.1 PyYAML-6.0 certifi-2023.5.7 charset-normalizer-2.0.12 contextvars-2.4 distro-1.8.0 idna-3.4 immutables-0.19 jmespath-0.10.0 looseversion-1.2.0 msgpack-1.0.5 psutil-5.9.5 pycryptodomex-3.18.0 pyzmq-25.0.2 requests-2.27.1 salt-3006.1 typing-extensions-4.1.1 urllib3-1.26.16

# salt-call --version
salt-call 3006.1

# salt-call --local grains.get saltversion
local:
    3006.1

# salt-call --local salt_version.equal Sulfur
local:
    False

# salt-call --local salt_version.get_release_number Sulfur
local:
    No version assigned.
```

### Previous Behavior
As mentioned above, the `salt_version.get_release_number` and `salt_version.equal` execution module functions (and probably others) are not working fine.

Other codenames seems working fine:

```console
# salt-call --local salt_version.get_release_number Neon
local:
    3000
```

After some debugging, I've noticed the following situations:
- Codename definitions for Salt >= 3006 seems wrong at `salt/version.py`, as they map the version codename to the `300X.0` release but not to the entire `300X`. This causes problems on codename detection.
- There seems to be a mess with the usage of `can_have_dot_zero` function that causes wrong a detection of Salt versions.

### New Behavior
Salt version resolution works fine now and versions are detected as expected:

```console
# salt-call --local salt_version.get_release_number Sulfur
local:
    3006

# salt-call --local salt_version.equal Sulfur
local:
    True
```

### Merge requirements satisfied?
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

meaksh committed:
Fix detection of Salt codename by salt_version module

meaksh committed:
Fix mess with version detection bad version definition

meaksh committed:
Add changes suggested by pre-commit

meaksh committed:
Add some new and fix unit tests

meaksh committed:
Add changelog file

meaksh committed:
Fix SaltStackVersion string for new versions format

s0undt3ch commented on tests/pytests/unit/modules/test_salt_version.py:
```diff
@@ -21,7 +21,7 @@ def test_mocked_objects():
         assert k == k.lower()
         assert isinstance(v, tuple)
-        if sv.new_version(major=v[0]) and not sv.can_have_dot_zero(major=v[0]):
+        if sv.new_version(major=v[0]):
```
So, from 3006 onward, the `.0` must always be present, the existing logic was correct, but SaltStackVersion is now reporting it differently?

s0undt3ch replied:
Ok, I see why this had to change.
And it looks like it hasn't broken other functionality.
```
❯ python
Python 3.10.2 (main, Nov 15 2022, 11:41:37) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import salt.version
>>> v = salt.version.SaltStackVersion.parse("3006")
>>> str(v)
'3006.0'
>>> v.info
v.info
>>> v.info
(3006, 0)
>>> v = salt.version.SaltStackVersion.parse("3005")
>>> str(v)
'3005'
>>> v.info
(3005,)
>>>
```

s0undt3ch commented on tests/pytests/unit/modules/test_salt_version.py:
```diff
@@ -64,6 +64,13 @@ def test_get_release_number_success_new_version():
+    """
+    Test that a version is returned for new versioning (3006)
+    """
+    assert salt_version.get_release_number("Sulfur") == "3006"
```
```suggestion
    assert salt_version.get_release_number("Sulfur") == "3006.0"
```
In theory, this should be the expected outcome... Although, all 3006 minor versions are Sulfur versions...

s0undt3ch commented on changelog/64554.fixed.md:
This file should be named `64306.fixed.md`

meaksh replied:
Done. Thanks for the review!

meaksh committed:
Rename changelog file according to bug report

s0undt3ch commented:
@meaksh since you've fixed most of the issues, could you also add a test case for the error shown in https://github.com/saltstack/salt/issues/64306 ?
Let me know if you don't have the bandwidth and I can push it to your PR.

meaksh commented:
@s0undt3ch ok, I'll give it a try to implement a test that covers the error shown at https://github.com/saltstack/salt/issues/64306

meaksh committed:
Do not crash when passing numbers to 'salt_version.get_release_number'

meaksh commented:
@s0undt3ch I've added the mentioned test.

I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in https://github.com/saltstack/salt/issues/64306).

I've just adjusted to the code to raise a `CommandExecutionError` instead.

s0undt3ch commented:
> @s0undt3ch I've added the mentioned test.
> 
> I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in #64306).
> 
> I've just adjusted to the code to raise a `CommandExecutionError` instead.

Thanks!

Ch3LL reviewed (changes requested):
The documentation for `get_release_number` states it should return `MAJOR.PATCH` but your PR is only returning `MAJOR` for 3006 which now does include a patch in the version. The expected behavior seems to be to return `3006.0` instead of `3006`.

ggiesen commented:
If the intention is to not 

> > @s0undt3ch I've added the mentioned test.
> > I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in #64306).
> > I've just adjusted to the code to raise a `CommandExecutionError` instead.
> 
> Thanks!

If the intention is not to fix this, then please update the example in the docs:

https://github.com/saltstack/salt/blob/9aeed74e96542558fc473a24d6b38add4a62b65b/salt/modules/salt_version.py#L18-L29

meaksh committed:
Fix salt_version execution module documentation

meaksh commented:
@Ch3LL as mentioned, the documentation for `salt_version` is now upgraded to reflect the expected behavior.

meaksh committed:
Fix test to get proper formatted version

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

s0undt3ch commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[<img src="https://img.shields.io/github/pulls/detail/state/saltstack/salt/64671">](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
